libdrgn/arch_ppc64.c

```diff
@@ -267,6 +280,193 @@ apply_elf_reloc_ppc64(const struct drgn_relocating_section *relocating,
+
+	/*
+	 * Only BOOK3S CPU family is supported.
+	 * If interrupt_base_book3e is missing then it is BOOK3S.
+	 */
+	err = drgn_program_find_object(prog, "interrupt_base_book3e", NULL,
+				       DRGN_FIND_OBJECT_ANY, &book3s);
+	if (!err) {
+		err = drgn_error_create(DRGN_ERROR_OTHER,
+					"Virtual address translation is not available for BOOK3E CPU family.");
+		goto out;
+	}
+
+	/* Identify the MMU type. */
+	err = drgn_program_find_object(prog, "cur_cpu_spec", NULL,
+				       DRGN_FIND_OBJECT_ANY, &cur_cpu_spec);
+	if (err)
+		goto out;
+	err = drgn_object_member_dereference(&mmu_features_obj, &cur_cpu_spec,
+					     "mmu_features");
+	if (err)
+		goto out;
+
+	err = drgn_object_read_unsigned(&mmu_features_obj, &mmu_features);
+
+	if (err)
+		goto out;
+
+	if (!(mmu_features & RADIX_MMU)) {
+		err = drgn_error_create(DRGN_ERROR_OTHER,
+					"Virtual address translation is only supported for Radix MMU.");
+		goto out;
+	}
+
+	if (page_size == PAGE_64K)
+		it->pt_levels = pt_levels_radix_64k;
+	else
+		it->pt_levels = pt_levels_radix_4k;
+
+	*ret = &it->it;
+
+out:
+	drgn_object_deinit(&book3s);
+	drgn_object_deinit(&cur_cpu_spec);
+	drgn_object_deinit(&mmu_features_obj);
+	return err;
+}
+
+static void
+linux_kernel_pgtable_iterator_destroy_ppc64(struct pgtable_iterator *_it)
+{
+	struct pgtable_iterator_ppc64 *it =
+		container_of(_it, struct pgtable_iterator_ppc64, it);
+	free(it);
+}
+
+static void
+linux_kernel_pgtable_iterator_init_ppc64(struct drgn_program *prog,
+					       struct pgtable_iterator *_it)
+{
+	return;
+}
+
+static struct drgn_error *
+linux_kernel_pgtable_iterator_next_ppc64(struct drgn_program *prog,
+					 struct pgtable_iterator *_it,
+					 uint64_t *virt_addr_ret,
+					 uint64_t *phys_addr_ret)
+{
+	uint16_t level;
+	uint64_t virt_addr;
+	struct drgn_error *err;
+	struct pgtable_iterator_ppc64 *it;
+	const bool bswap = drgn_platform_is_little_endian(&prog->platform);
+
+	it = container_of(_it, struct pgtable_iterator_ppc64, it);
+	level = it->levels;
+	virt_addr = it->it.virt_addr;
+
+	for (;;level--) {
+		uint64_t index;
+		uint64_t table;
+		uint64_t entry;
+		bool table_physical;
+
+		if (level == it->levels) {
+			table = it->it.pgtable;
+			table_physical = false;
+		} else {
+			if (bswap)
+				entry = bswap_64(entry);
+
+			/* PAGE_PTE bit represents huge page */
+			if (entry & PAGE_PTE || level == 0) {
+				uint64_t mask = (UINT64_C(1) << it->pt_levels[level].shift) - 1;
+				*virt_addr_ret = virt_addr & ~mask;
+				if (level == 0 && !(entry & PAGE_PRESENT))
+					*phys_addr_ret = UINT64_MAX;
+				else
+					*phys_addr_ret = entry & get_page_mask(it, level);
+
+				it->it.virt_addr = (virt_addr | mask) + 1;
+				return NULL;
+			}
+
+			table = entry & ~it->pt_mask;
+			table_physical = true;
+		}
+
+		index = get_index(it, virt_addr, level);
+		err = drgn_program_read_memory(prog, &entry,
+					       table + (8 * index), 8,
+					       table_physical);
+		if (err)
+			return err;
```

Comment:
Nice. I will use drgn_program_read_u64 API to read a entry.

Reply:
@osandov I agree that reading reading 64 bit using `drgn_program_read_u64` is better compare to `drgn_program_read_memory` but how come `drgn_program_read_u64` function knows page entries are stored in big endian?

Reply:
The definition of `drgn_program_read_u64()` comes from this macro: https://github.com/osandov/drgn/blob/62bd77ca532ec989e2b5b450509544e0ed80b499/libdrgn/program.c#L1724-L1746

That calls `drgn_program_bswap()` to check whether the byte order is different than the byte order of the host:
https://github.com/osandov/drgn/blob/62bd77ca532ec989e2b5b450509544e0ed80b499/libdrgn/program.h#L258-L271 https://github.com/osandov/drgn/blob/62bd77ca532ec989e2b5b450509544e0ed80b499/libdrgn/platform.h#L455-L464

The byte order of the target is determined from the core dump metadata, either from the ELF header:
https://github.com/osandov/drgn/blob/62bd77ca532ec989e2b5b450509544e0ed80b499/libdrgn/platform.c#L186-L187

Or kdump attributes:
https://github.com/osandov/drgn/blob/62bd77ca532ec989e2b5b450509544e0ed80b499/libdrgn/kdump.c#L54-L62

Reply:
I agree, but I think I still need to swap the entry because even though the target platform is little-endian, the page entries are stored in big-endian.

However, using drgn_program_read_u64 along with bswap if the target platform is little-endian brings the entry to the correct order for all four LE/BE combinations.

Reply:
Oh, so even on ppc64le, the page table entries are big-endian?

Reply:
Yes.

Reply:
Ok, I'm sorry, I misunderstood. I thought they matched the byte order of the target. In that case, the best thing to do would be to go back to using `drgn_program_read_memory()` and use `be64toh()`, with something like:

```c
#include <endian.h>

uint64_t entry;
err = drgn_program_read_memory(prog, &entry, table + (8 * index), 8, table_physical);
if (err)
	return err;
// Page table entries are always big endian.
entry = be64toh(entry);
```

Reply:
Sure let me update the patch.